src/Compiler/TypedTree/TypedTreeOps.fs

```diff
@@ -10095,14 +10135,549 @@ let (|CompiledForEachExpr|_|) g expr =
+    | Expr.Const (value = Const.Int32 start), Expr.Const (value = Const.Int32 step), Expr.Const (value = Const.Int32 finish) when finish < start && step > 0 || finish > start && step < 0 -> ValueSome EmptyRange
+    | Expr.Const (value = Const.Int64 start), Expr.Const (value = Const.Int64 step), Expr.Const (value = Const.Int64 finish) when finish < start && step > 0L || finish > start && step < 0L -> ValueSome EmptyRange
+    | Expr.Const (value = Const.UInt64 start), Expr.Const (value = Const.UInt64 _), Expr.Const (value = Const.UInt64 finish) when finish < start -> ValueSome EmptyRange
+    | Expr.Const (value = Const.UInt32 start), Expr.Const (value = Const.UInt32 _), Expr.Const (value = Const.UInt32 finish) when finish < start -> ValueSome EmptyRange
+
+    // sizeof<nativeint> is not constant, so |𝑐| ≥ 0x80000000n cannot be treated as a constant.
+    | Expr.Const (value = Const.IntPtr start), Expr.Const (value = Const.IntPtr step), Expr.Const (value = Const.IntPtr finish) when
+        uint64 start < 0x80000000UL
+        && uint64 step < 0x80000000UL
+        && uint64 finish < 0x80000000UL
+        && (finish < start && step > 0L || finish > start && step < 0L)
+        ->
+        ValueSome EmptyRange
+
+    // sizeof<unativeint> is not constant, so |𝑐| > 0xffffffffun cannot be treated as a constant.
+    | Expr.Const (value = Const.UIntPtr start), Expr.Const (value = Const.UIntPtr step), Expr.Const (value = Const.UIntPtr finish) when
+        start <= 0xffffffffUL
+        && step <= 0xffffffffUL
+        && finish <= 0xffffffffUL
+        && finish <= start
+        ->
+        ValueSome EmptyRange
+
+    | Expr.Const (value = Const.Int16 start), Expr.Const (value = Const.Int16 step), Expr.Const (value = Const.Int16 finish) when finish < start && step > 0s || finish > start && step < 0s -> ValueSome EmptyRange
+    | Expr.Const (value = Const.UInt16 start), Expr.Const (value = Const.UInt16 _), Expr.Const (value = Const.UInt16 finish) when finish < start -> ValueSome EmptyRange
+    | Expr.Const (value = Const.SByte start), Expr.Const (value = Const.SByte step), Expr.Const (value = Const.SByte finish) when finish < start && step > 0y || finish > start && step < 0y -> ValueSome EmptyRange
+    | Expr.Const (value = Const.Byte start), Expr.Const (value = Const.Byte _), Expr.Const (value = Const.Byte finish) when finish < start -> ValueSome EmptyRange
+    | Expr.Const (value = Const.Char start), Expr.Const (value = Const.Char _), Expr.Const (value = Const.Char finish) when finish < start -> ValueSome EmptyRange
+    | _ -> ValueNone
+
+/// Note: this assumes that an empty range has already been checked for
+/// (otherwise the conversion operations here might overflow).
+[<return: Struct>]
+let (|ConstCount|_|) (start, step, finish) =
+    match start, step, finish with
+    // Such a range would use up the entire address space, but for parity with the library implementation we need it to fail at runtime, not at build time.
```

Comment:
This is a dubious parity but I get it :)

Out of curiosity (not for this PR or whatever) - what do you think would be the effort to switch both things to fail at the build time?

Reply:
- It would be trivial to raise a compile-time error here for this case and for when `step` is a 0 literal.
- It seems like it would probably be straightforward to do that when type-checking application of the library function, too.

Raising an error at compile time for a constant 0 step seems pretty harmless, since such code always currently fails at runtime anyway.

Raising a build-time error for potentially-too-big ranges has some nuances, though:

- A `for`-loop over the range $0..2^{64}$ might take a while, but it's not strictly _wrong_, so we wouldn't want to raise a build-time error when $count$ was used for that.
- When $count$ is being used to initialize an array, [`System.Array.MaxLength`](https://learn.microsoft.com/dotnet/api/system.array.maxlength?view=net-8.0) is technically the limit (and is not a constant).
- Initializing a list with $count \geq 2^{32}$ is theoretically fine on a 64-bit machine with sufficient memory, but not if the same code is run on x86.

Reply:
Alright, thanks for the research and the explanation!